File: demo/typescript-service/src/after/user-service.ts
// Refactored UserService with improved architecture
import { validateUserData } from './validation.js';
import { NotificationService } from './notifications.js';
import { ThemeService } from './theme.js';
import { Logger } from './logger.js';

export class UserService {
  private users: Map<string, any> = new Map();

  constructor(
    private logger: Logger,
    private notificationService: NotificationService,
    private themeService: ThemeService
  ) {}

  async processUser(userData: any): Promise<void> {
    // Early validation with clear error messages
    validateUserData(userData);
    
    // Log the operation
    this.logger.log(`Processing user ${userData.id}`);
    
    // Store user
    this.users.set(userData.id, userData);
    
    // Handle notifications if preferences exist
    this.handleNotifications(userData);
    
    // Handle theme preferences
    this.handleThemePreferences(userData);
    
    // Audit logging for admin users
    if (userData.role === 'admin') {
      this.logger.logAudit(`Admin user created: ${userData.id}`);
    }
    
    // Track analytics
    this.logger.logAnalytics(userData);
    
    console.log(`User ${userData.id} processed successfully`);
  }

  private handleNotifications(userData: any): void {
    if (!userData.preferences?.notifications) return;
    
    const { notifications } = userData.preferences;
    
    if (notifications.email) {
      this.notificationService.sendEmail(userData.email, 'Welcome!');
    }
    
    if (notifications.sms && userData.phone) {
      this.notificationService.sendSMS(userData.phone, 'Welcome!');
    }
  }

  private handleThemePreferences(userData: any): void {
    const theme = userData.preferences?.theme;
    if (theme) {
      this.themeService.applyTheme(userData.id, theme);
    }
  }

  public getUser(id: string): any | undefined {
    return this.users.get(id);
  }

  public getAllUsers(): any[] {
    return Array.from(this.users.values());
  }
}

File: demo/csharp-library/src/after/UserService.cs
// Refactored UserService with improved architecture
using System;
using System.Collections.Generic;
using System.IO;
using UserManagement.Validation;
using UserManagement.Services;

namespace UserManagement
{
    public class UserService
    {
        private readonly Dictionary<string, User> users = new Dictionary<string, User>();
        private readonly ILogger logger;
        private readonly INotificationService notificationService;
        private readonly IThemeService themeService;

        public UserService(
            ILogger logger,
            INotificationService notificationService,
            IThemeService themeService)
        {
            this.logger = logger;
            this.notificationService = notificationService;
            this.themeService = themeService;
        }

        public void ProcessUser(User userData)
        {
            // Early validation with clear error messages
            UserValidator.ValidateUser(userData);
            
            // Log the operation
            logger.Log($"Processing user {userData.Id}");
            
            // Store user
            users[userData.Id] = userData;
            
            // Handle notifications
            HandleNotifications(userData);
            
            // Handle theme preferences
            HandleThemePreferences(userData);
            
            // Audit logging for admin users
            if (userData.Role == "Admin")
            {
                logger.LogAudit($"Admin user created: {userData.Id}");
            }
            
            // Track analytics
            logger.LogAnalytics($"User created: {userData.Id}");
            
            Console.WriteLine($"User {userData.Id} processed successfully");
        }

        private void HandleNotifications(User userData)
        {
            if (userData.Preferences?.Notifications == null) return;
            
            var notifications = userData.Preferences.Notifications;
            
            if (notifications.Email)
            {
                notificationService.SendEmail(userData.Email, "Welcome!");
            }
            
            if (notifications.SMS && !string.IsNullOrEmpty(userData.Phone))
            {
                notificationService.SendSMS(userData.Phone, "Welcome!");
            }
        }

        private void HandleThemePreferences(User userData)
        {
            var theme = userData.Preferences?.Theme;
            if (!string.IsNullOrEmpty(theme))
            {
                themeService.ApplyTheme(userData.Id, theme);
            }
        }

        public User GetUser(string id) => users.TryGetValue(id, out var user) ? user : null;

        public IEnumerable<User> GetAllUsers() => users.Values;
    }

    // Logger interface
    public interface ILogger
    {
        void Log(string message);
        void LogAudit(string message);
        void LogAnalytics(string data);
    }

    // Theme service interface
    public interface IThemeService
    {
        void ApplyTheme(string userId, string theme);
    }
}
